agent/lib/kontena/observer.rb

```diff
@@ -43,75 +93,382 @@ def values
+            task.resume(self)
+          else
+            fail "observe is not active"
+          end
+        end
+
+        # No longer expecting updates.
+        #
+        def kill
+          @alive = false
+        end
       end
     end
 
-    # Observe has been updated, call if ready.
-    # Called from observe as an async task.
+    # Register handler for Kontena::Observable::Message on the Celluloid::Actor
     #
-    # @param observe [Observer::Observe]
-    def observed(observe)
-      observe.call if observe.ready?
+    def register_observer_handler
+      actor = Thread.current[:celluloid_actor]
+      actor.observe_handler ||= actor.handle(Kontena::Observable::Message) do |message|
+        # This handler runs directly in the Actor thread, outside of any task context.
+        handle_observer_message(message)
+      end
     end
 
-    # Called from Observable as an async task.
+    # Update the Observe, and call it if ready, which will active the observe task.
+    # This must run directly in the Actor thread, outside of any task context.
+    # Called from Celluloid::Actor#handle or Celluloid::Actor@timers.after
     #
-    # @param observe [Observer::Observe]
-    # @param observable [Celluloid::Proxy::Cell<Observable>]
-    # @param value [Object, nil] observed value
-    def update_observe(observe, observable, value)
-      debug "observe Observable<#{observable.__klass__}> -> #{value}"
-
-      observe.set(observable, value)
-      observed(observe)
+    # @param message [Kontena::Observable::Message]
+    def handle_observer_message(message)
+      observe = message.observe
+
+      if message.observable
+        debug "observe update #{message.describe_observable} -> #{message.value}"
+
+        observe.set(message.observable, message.value)
+      end
+
+      if !observe.alive?
+        debug "observe dead: #{observe.describe_observables}"
+      elsif !observe.active?
+        debug "observe paused: #{observe.describe_observables}"
+      elsif observe.ready?
+        debug "observe ready: #{observe.describe_observables}"
+        observe.call
+      else
+        debug "observe blocked: #{observe.describe_observables}"
+      end
     end
 
-    # Observe values from Observables, yielding each value to block.
+    # Block thread until mailbox receives update or times out.
+    # Must only be called in Celluloid.exclusive? mode.
+    # Returns once the observe is ready.
+    #
+    # @param observe [Observe]
+    # @param timeout [Float, nil]
+    # @raise Celluloid::TaskTimeout
+    def wait_observe(observe, mailbox, timeout: nil)
+      deadline = Time.now + timeout if timeout
+
+      while !observe.ready?
+        receive_timeout = deadline ? deadline - Time.now : nil
+
+        if receive_timeout && receive_timeout < 0
+          raise Celluloid::TaskTimeout.new("wait deadline exceeded")
+        end
+
+        message = mailbox.receive(receive_timeout) { |msg|
+          debug "observe receive #{msg.class.name}"
```

Comment:
OTOH this blocking mailbox receive should match how the existing `wait_until! { Celluloid::Proxy::Cell.ready? }` calls are implemented in exclusive mode:

* [`Celluloid::Proxy::Sync#method_missing`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid/proxy/sync.rb#L20-L22)
* [`Celluloid::Call::Sync#value`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid/call/sync.rb#L44)
* [`Celluloid::Call::Sync#response`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid/call/sync.rb#L41)
* [`Celluloid.suspend`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid.rb#L141)
* [`Celluloid::Call::Sync#wait`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid/call/sync.rb#L51)

So I'd be surprised if this somehow performed any worse... I guess it's just the debug logging that made it chew up so much CPU?